[Chng Xian Lin] iP

xianlinc opened the pull request:
# Skeltal
<blockquote>"“A man is but the product of his thoughts. What he thinks, he becomes.” - Gandhi</blockquote>

## Skeltal frees your mind of having to remember what tasks to do and when to do it

### It can help you:

- Manage your tasks
- Remember deadlines
- Save your tasks in a file
- Save lots of time


### These are the benefits you will gain as a programmer if you look at our code: 
1. Learn *Java* &#128507;
2. Learn good coding practices
3. Learn from our coding standards
#### You can download our code <a href="https://github.com/xianlinc/ip">here</a>
### Skeltal is:
- [x] Nice interface
- [x] Easy to use
- [x] Made by NUS
- [x] Completed product
### If you are a java programmer, heres the `main` method!
```java
public class Main {
    public static void main(String[] args) {
        Skeltal.run();
    }
}
```

Commit:
Add Gradle support

xianlinc committed:
Add greet, echo and exit functionality for chatbot

xianlinc committed:
Add storage functionality and refactor code

xianlinc committed:
Add checklist function

xianlinc committed:
Add polymorphism and Task class

xianlinc committed:
Add exception handling for user input

xianlinc committed:
Add delete task function

xianlinc committed:
Add testing unit

xianlinc committed:
Add ability to load task and save tasks

Skeltal is unable to save task lists
Add ability to store task list on any changes and load task list on startup

xianlinc committed:
Add parsing of dates for tasks

xianlinc committed:
Merge branch 'branch-Level-7'

xianlinc committed:
Merge branch 'branch-Level-8'

xianlinc committed:
Improve OOP in project

Add Ui and Parser classes.
Rename Manager Class to TaskList

xianlinc committed:
Refactor into skeltal package

xianlinc committed:
Add Junit testing

xianlinc committed:
Add javadocs to project

xianlinc committed:
Refactor code to follow coding standard

Code now follows camelCase

xianlinc committed:
Add find task function

xianlinc committed:
Merge branch 'branch-A-JavaDoc'

xianlinc committed:
Merge branch 'branch-A-CodingStandard'

xianlinc committed:
Merge branch 'branch-Level-9'

xlzior commented on src/main/java/skeltal/Deadline.java:
```diff
@@ -0,0 +1,60 @@
+     */
+    public Deadline(String rawTime) throws SkeltalException {
+        super(rawTime.split("/", 2)[0]);
+        String[] procTime = rawTime.split("/", 2);
```
What does `procTime` stand for?

xlzior commented on src/main/java/skeltal/Ui.java:
```diff
@@ -0,0 +1,39 @@
+    /**
+     * Prints an introduction statement.
+     */
+    public static void introduction() {
```
Perhaps this should be renamed to `introduce()`? According to the [guidelines on naming](https://nus-cs2103-ay2122s1.github.io/website/se-book-adapted/chapters/codeQuality.html#use-nouns-for-things-and-verbs-for-actions)

xlzior commented on src/main/java/skeltal/Storage.java:
```diff
@@ -0,0 +1,79 @@
+            Scanner sc =  new Scanner(taskFile);
+            while (sc.hasNextLine()) {
+                String data = sc.nextLine();
+                String[] dataPr = data.split(" \\| ");
```
What does `dataPr` stand for?

xlzior commented on src/main/java/skeltal/Skeltal.java:
```diff
@@ -0,0 +1,29 @@
+        TaskList.loadTaskList(Storage.loadFile());
+        Ui.introduction();
+        Scanner sc = new Scanner(System.in);
+        boolean shutdown = false;
```
Perhaps this can be renamed to sound more like a boolean? According to the [naming guidelines](https://se-education.org/guides/conventions/java/intermediate.html#naming)

xlzior commented on src/main/java/skeltal/TaskList.java:
```diff
@@ -0,0 +1,91 @@
+     * A method to add a Task object into the list.
+     * @param assignment A task object.
+     */
+    public static void addReply(Task assignment) {
```
Any reason why it's named `addReply`? It seems like you're adding a reply rather than adding a task 😅

xlzior commented on src/main/java/skeltal/TaskList.java:
```diff
@@ -0,0 +1,91 @@
+     * Sets the Task at the index (index_1 - 1) in the TaskList to a "completed" state.
+     * @param index_1 The index of the task in the list + 1.
+     */
+    public static void done(String index_1) {
```
It seems like you named this `index_1` since it represents `index + 1` rather than `index`, but is there another way to make this differentiation? Variable names should be in camelCase ([naming](https://se-education.org/guides/conventions/java/intermediate.html#naming)).

xlzior commented on src/main/java/skeltal/TaskList.java:
```diff
@@ -0,0 +1,91 @@
+     * Returns a String that represents the loadable form of the TaskList.
+     * @return A String representation of the TaskList
+     */
+    public static String storeTasks() {
```
I like your function names and JavaDocs! very clear

junghon3709 commented on src/main/java/skeltal/Parser.java:
```diff
@@ -0,0 +1,67 @@
+            String[] arrReply = processReply(userReply);
+            System.out.println(Ui.line());
+            switch(arrReply[0]) {
+                case "list":
```
I believe according to the coding standard, the case should be aligned with switch? You can configure this behaviour in Intellj here: [https://stackoverflow.com/questions/40888798/config-intellij-switch-case-style#:~:text=1%20Answer&text=Go%20to%20Settings%20%3E%20Editor%20%3E%20Code,Indent%20'case'%20branches%22.](url)

junghon3709 commented on src/main/java/skeltal/Event.java:
```diff
@@ -0,0 +1,61 @@
+     * @return A String representation of the Event object that is readable by the loader.
+     */
+    @Override
+    public String store() {
```
You might want to ban users from entering "|".

junghon3709 commented on src/main/java/skeltal/Ui.java:
```diff
@@ -0,0 +1,39 @@
+ */
+public class Ui {
+
+    private static String line = "---------------------------------------------";
```
Maybe you should put `final` and line as LINE since it is a constant.

junghon3709 commented on src/test/java/SkeltalTest.java:
```diff
@@ -0,0 +1,37 @@
+import java.io.ByteArrayOutputStream;
+import java.io.PrintStream;
+
+import org.junit.jupiter.api.BeforeEach;
```
I think the imports should be in alphabetical order according to the style guide. Not sure if it's mentioned in the Java coding standard, but you will run into problems in checkstyle if you do implement it.

junghon3709 commented on src/test/java/SkeltalTest.java:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    @Test
+    public void testParser() {
```
Your name for tests should also follow a specific style: see [https://se-education.org/guides/conventions/java/intermediate.html#naming](link).

junghon3709 commented on src/test/java/SkeltalTest.java:
```diff
@@ -0,0 +1,37 @@
+    @Test
+    public void testParser() {
+        Parser.response("todo quick");
+        assertEquals("---------------------------------------------\n" +
```
I believe the `+` signs should be at the start of line? Not the end. It's not explictly stated here: [https://se-education.org/guides/conventions/java/intermediate.html#layout](url), but you run into problems when implementing checkstyle later. You can configure Intellj to do this behaviour here: [https://stackoverflow.com/questions/6408462/intellij-idea-plus-sign-when-string-wrap](url)

junghon3709 commented on src/main/java/skeltal/Parser.java:
```diff
@@ -0,0 +1,67 @@
+     */
+    public static void response(String userReply) {
+        try {
+            String[] arrReply = processReply(userReply);
```
Maybe you should rename it to arrayReplies to signify plural

junghon3709 commented on src/main/java/skeltal/Storage.java:
```diff
@@ -0,0 +1,79 @@
+     * @return An ArrayList of Tasks representative of the file.
+     */
+    public static ArrayList<Task> loadFile() {
+        ArrayList<Task> arrayList = new ArrayList<>();
```
Maybe name `arrayList` into something more descriptive, like `taskList`?

junghon3709 commented on src/main/java/skeltal/Storage.java:
```diff
@@ -0,0 +1,79 @@
+                        arrayList.add(todo);
+                        break;
+                    case "E":
+                        Event event = new Event(dataPr[2] +" /" + dataPr[3]);
```
`+` should have right spacing?

junghon3709 commented on src/main/java/skeltal/Ui.java:
```diff
@@ -0,0 +1,39 @@
+     * Returns a line that is used to enclose Skeltal's responses.
+     * @return A String of a line.
+     */
+    public static String line() {
```
Maybe you could rename this to something more descriptive, such as `getResponse`.

junghon3709 reviewed (commented):
Clean code overall! Just some minor fixes :)

xianlinc committed:
Merge remote-tracking branch 'origin/add-gradle-support'

xianlinc committed:
Add gradle support

xianlinc committed:
Add GUI to skeltal

Previously there was no GUI and no simple way to style our chatbot.
Now with FXML and a GUI run using javafx, it will be easier for users to interact.

xianlinc committed:
Merge branch 'branch-Level-10'

xianlinc committed:
Add assertions to project

There are several assumptions that we want to make sure are valid throughout the running of the program.

Adding assertions will allow us to make sure those assumptions are still valid.

let's:
* use assertions to make sure variables are as expected
* make sure that the parsing of userInputs does not create null objects

xianlinc committed:
Enforce code quality

There are several classes and functions that have been made redundant due to the GUI.

Code quality can be improved if we refactor and remove those classes.

Let's remove and refactor the classes so that we have cleaner more legible code.

xianlinc committed:
Merge pull request #2 from xianlinc/branch-A-Assertions

Add assertions to project

xianlinc committed:
Add SLAP to parser class

Parser class has many operations being done in the switch statement.

It will increase legibility if we add more SLAP to the class.

Add SLAP to the parser class.

xianlinc committed:
Resolve merge conflicts into origin

xianlinc committed:
Merge pull request #3 from xianlinc/branch-A-CodeQuality

A-CodeQuality

xianlinc committed:
Remove Skeltal.java and fix bug with intro

xianlinc committed:
Add ability to handle expenses

Expenses are not supported in the current version of Skeltal.

It will be useful to be able to add, delete, sum and list expenses using Skeltal.

Let's:
*Add ability to add expenses
*Add ability to delete expenses
*Add ability to list expenses
*Add ability to sum expenses

xianlinc committed:
Merge branch 'D-Expenses'

xianlinc committed:
Refactor classes into packages

The project has too many classes that are not in packages.

It will be easier to understand if there were more packages to categorise the classes.

Let's refactor the classes in this structure:
*skeltal:
  *task
    *expense
    *types
*skeltalgui

xianlinc committed:
Improve Javadocs

xianlinc committed:
Add checkstyle

xianlinc committed:
Merge pull request #5 from xianlinc/A-CheckStyle

Add checkstyle

xianlinc committed:
Update test module

xianlinc committed:
Improve GUI

xianlinc committed:
Merge pull request #6 from xianlinc/branch-A-BetterGui

Improve GUI

xianlinc committed:
Add website to project

xianlinc committed:
Update Readme.md

xianlinc committed:
bug

xianlinc committed:
Fix bug where jar file cannot run

skeltal.jar cannot run on other systems.

Making sure that the javafx library is added to the jar file will fix the issue.

Let's change build.gradle to include javafx for all systems.

xianlinc committed:
Merge remote-tracking branch 'origin/A-bug' into bug-fix

xianlinc committed:
Merge branch 'bug-fix'

xianlinc committed:
Fix formatting error of minus method

xianlinc committed:
Update docs

Update Readme.md and add Ui.png to docs.

xianlinc committed:
Move docs to correct folder

xianlinc committed:
Fix formatting error

##Feature is supposed to be ## Feature
